File: 2024/elixir/day02.exs
# https://adventofcode.com/2024/day/2

defmodule Day02 do
  def part1(input) do
    input
    |> parse()
    |> Enum.count(&safe?/1)
  end

  def part2(input) do
    input
    |> parse()
    |> Enum.count(fn report ->
      safe?(report) or Enum.any?(dampened(report), &safe?/1)
    end)
  end

  defp dampened(report) do
    0..length(report)
    |> Enum.map(&List.delete_at(report, &1))
  end

  defp safe?(report) do
    processed =
      for [left, right] <- Enum.chunk_every(report, 2, 1) do
        diff = abs(left - right)

        cond do
          diff == 0 or diff > 3 -> nil
          left > right -> :inc
          right > left -> :dec
        end
      end

    case Enum.uniq(processed) do
      [v] when v in [:inc, :dec] -> true
      _ -> false
    end
  end

  defp parse(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      String.split(line, " ") |> Enum.map(&String.to_integer/1)
    end)
  end
end

"""
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""
|> then(fn input ->
  {
    Day02.part1(input),
    Day02.part2(input)
  }
  |> IO.inspect()
end)

File: 2024/ruby/day02.rb
#!/usr/bin/env ruby

def safe?(report)
  inc = nil

  report.each_cons(2) { |(left, right)|
    return false unless (1..3).include?((left - right).abs)

    if inc.nil?
      inc = left > right
    else
      return false if inc != (left > right)
    end
  }

  true
end

def dampened(report)
  report.size.times.map { |n|
    report.dup.tap { _1.tap { |r| r.delete_at(n) } }
  }
end

reports = DATA.readlines.map { _1.scan(/\d+/).map(&:to_i) }
reports.count { safe?(_1) }.then { puts "P1: #{_1}" }
reports.count { safe?(_1) or dampened(_1).any?(&method(:safe?)) }.then { puts "P2: #{_1}" }

__END__
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
